src/services/serviceProvider.ts

```diff
@@ -51,6 +56,113 @@ export class ServiceProvider {
     instanceName?: string,
     ...rest: ServiceParams<T>[]
   ): Promise<ServiceReturnType<T>> {
+    const { serviceInstance, correctedInstanceName } =
+      ServiceProvider.resolveServiceInstance(type, instanceName);
+
+    if (!serviceInstance) {
+      const command = ServiceProvider.getCommandString(type);
+      await ServiceProvider.checkCommandAvailability(command);
+      return ServiceProvider.materializeService<T>(
+        type,
+        correctedInstanceName,
+        ...rest
+      );
+    }
+
+    return serviceInstance;
+  }
+
+  /**
+   * Retrieves a service instance of the specified type and instance name, waiting for the service to become available if necessary.
+   * If the service instance does not exist, it will be created.
+   * @template T - The type of the service.
+   * @param {T} type - The type of the service.
+   * @param {string} [instanceName] - The name of the service instance.
+   * @param {WaitOptions} [options] - Options for waiting for the service to become available.
+   * @param {...ServiceParams<T>} parameters - Additional parameters for the service.
+   * @returns {Promise<ServiceReturnType<T>>} The service instance.
+   */
+  static async getServiceWithWait<T extends ServiceType>(
+    type: T,
+    instanceName?: string,
+    options?: WaitOptions,
+    ...parameters: ServiceParams<T>[]
+  ): Promise<ServiceGetResult<T>>;
+
+  /**
+   * Retrieves a service instance of the specified type and instance name, waiting for the service to become available if necessary.
+   * If the service instance does not exist, it will be created.
+   * @template T - The type of the service.
+   * @param {T} type - The type of the service.
+   * @param {string} [instanceName] - The name of the service instance.
+   * @param {...ServiceParams<T>} parameters - Additional parameters for the service.
+   * @returns {Promise<ServiceReturnType<T>>} The service instance.
+   */
+  static async getServiceWithWait<T extends ServiceType>(
+    type: T,
+    instanceName?: string,
+    ...parameters: ServiceParams<T>[]
+  ): Promise<ServiceGetResult<T>>;
+
+  /**
+   * Retrieves a service instance of the specified type and instance name, waiting for the service to become available if necessary.
+   * If the service instance does not exist, it will be created.
+   * @template T - The type of the service.
+   * @param type - The type of the service.
+   * @param [instanceName] - The name of the service instance.
+   * @param [optionsOrParameter] - Options for waiting for the service to become available or additional parameters for the service.
+   * @param rest - Additional parameters for the service.
+   * @returns The service instance.
+   * @private
+   */
+  static async getServiceWithWait<T extends ServiceType>(
```

Comment:
Are there unit tests for this method as well or is this something that's normally added in E2E testing?

Reply:
unit tests are a todo for the new "stuff", mocking out vscode and other goodies when vscode is not really present is difficult.